### PR TITLE
dvc api

### PR DESCRIPTION
**Pull request by jckhang**
## What does this PR do?

add dvc read function
add api to get csv blob handle without dag and node
api to set csvblob as external owned.

## Changes

- [x] add dvc read
- [x] api to get csv blob without dag
- [x] api to set blob owner

## Before reviewing

-   [x] Are there any linked PRs?
-   [x] Did you test this PR?
-   [ ] Is there any migration or seeding necessary?
-   [ ] Did you add api to only one of the language (python/typescript)?

### Links:

-   JIRA: https://myaccern.atlassian.net/browse/TAD-2943
-  xyme-backend: https://github.com/Accern/xyme-backend/pull/136

### New API added that not implemented in other languages

If your answer is yes, please also update the todo_api.txt at where the API is missing.

### CODE REVIEW

File: packages/python/accern_xyme/accern_xyme.py [hunk 1]
```diff
@@ -834,6 +866,33 @@ def get_triton_models(self) -> List[str]:
...
+                github URL or a file system path.
+            rev (str):
+                Git commit (any revision such as a branch or tag name, or a
+                commit hash). If repo is not a Git repo, this option is
+                ignored. Default: HEAD.
+
+        Returns:
+            the content of the file.
+        """
+        import dvc.api
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:887]: try.....except ImportError install from .....

File: packages/python/accern_xyme/accern_xyme.py [hunk 2]
```diff
@@ -1788,32 +1855,43 @@ def fix_error(self) -> NodeState:
...
-        return JSONBlobHandle(self._client, res["json"], res["count"])
+                "key": blob_key,
+                "type": blob_type,
+            }))["uri"]
+
+    def get_csv_blob(self, key: str = "orig") -> 'CSVBlobHandle':
+        return CSVBlobHandle(
+            self._client,
+            self.get_blob_uri(key, "csv"),
+            self.as_owner())
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:1872]: cannot assume self as owner

File: packages/python/accern_xyme/accern_xyme.py [hunk 3]
```diff
@@ -622,6 +638,22 @@ def create_new_dag(
                 "index": index,
             }))["dag"]
 
+    def get_csv_blob(self, blob_uri: str) -> 'CSVBlobHandle':
+        return CSVBlobHandle(
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:642]: how could we verify that the type is correct?

File: packages/python/accern_xyme/accern_xyme.py (hunk 1 above)
Comment (JosuaKrause) [reply packages/python/accern_xyme/accern_xyme.py:887]: make fully optional -- for pylint put in requirements

File: packages/python/accern_xyme/accern_xyme.py [hunk 4]
```diff
@@ -1604,12 +1669,13 @@ def from_node_info(
         return res
 
     def update_info(self, node_info: NodeInfo) -> None:
-        if self._node_id != node_info["id"]:
+        if self.get_id() != node_info["id"]:
             raise ValueError(f"{self._node_id} != {node_info['id']}")
         self._node_name = node_info["name"]
         self._type = node_info["type"]
+        owner = self.as_owner()
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:1676]: can't assume owner

Comment (JosuaKrause) [reply packages/python/accern_xyme/accern_xyme.py:1676]: is owner really necessary?
can't we just lazily get the owner when we *actually* need them?

File: packages/python/accern_xyme/accern_xyme.py [hunk 5]
```diff
@@ -1964,12 +1996,14 @@ def __init__(
             self,
             client: XYMEClient,
             uri: str,
-            is_full: bool) -> None:
+            is_full: bool,
+            owner: BlobOwner) -> None:
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2000]: Optional

File: packages/python/accern_xyme/accern_xyme.py [hunk 6]
```diff
@@ -1964,12 +1996,14 @@ def __init__(
...
         self._client = client
         self._uri = uri
         self._is_full = is_full
         self._ctype: Optional[str] = None
         self._tmp_uri: Optional[str] = None
+        self._owner = owner
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2006]: `get_owner` calls `get_blob_owner` and fills in the cache if None

File: packages/python/accern_xyme/accern_xyme.py [hunk 7]
```diff
@@ -2017,7 +2051,7 @@ def list_files(self) -> List['BlobHandle']:
                 "blob": self.get_uri(),
             }))
         return [
-            BlobHandle(self._client, blob_uri, is_full=True)
+            BlobHandle(self._client, blob_uri, is_full=True, owner=self._owner)
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2054]: get_owner

File: packages/python/accern_xyme/accern_xyme.py [hunk 8]
```diff
@@ -2035,12 +2069,13 @@ def set_owner(self, owner: NodeHandle) -> BlobOwner:
             }))
 
     def get_owner(self) -> BlobOwner:
-        if self.is_full():
-            raise ValueError(f"URI must not be full: {self}")
-        return cast(BlobOwner, self._client.request_json(
-            METHOD_GET, "/blob_owner", {
-                "blob": self.get_uri(),
-            }))
+        return self._owner
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2072]: if self._owner is None:....

File: packages/python/accern_xyme/accern_xyme.py [hunk 9]
```diff
@@ -2035,12 +2069,13 @@ def set_owner(self, owner: NodeHandle) -> BlobOwner:
...
+
+    def get_owner_dag(self) -> Optional[str]:
+        return self._owner["owner_dag"]
+
+    def get_owner_node(self) -> str:
+        return self._owner["owner_node"]
 
     def copy_to(
             self,
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2081]: how could you make it an external owner?

File: packages/python/accern_xyme/accern_xyme.py [hunk 10]
```diff
@@ -2203,8 +2244,7 @@ def __init__(
             client: XYMEClient,
             uri: str,
             owner: BlobOwner) -> None:
-        super().__init__(client, uri, is_full=False)
-        self._owner = owner
+        super().__init__(client, uri, is_full=False, owner=owner)
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2247]: constructor needed?

File: packages/python/accern_xyme/accern_xyme.py [hunk 11]
```diff
@@ -2268,33 +2308,104 @@ def __init__(
...
+            METHOD_PUT, "/model_setup", {
+                "dag": self.get_owner_dag(),
+                "node": self.get_owner_node(),
+                "config": obj,
+            }))
+
+    def get_model_params(self) -> ModelParamsResponse:
+        return cast(ModelParamsResponse, self._client.request_json(
+            METHOD_GET, "/model_params", {
+                "dag": self.get_owner_dag(),
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2404]: why is that needed? it's read-only, no?

File: packages/python/accern_xyme/accern_xyme.py [hunk 12]
```diff
@@ -2352,7 +2463,6 @@ def __ne__(self, other: object) -> bool:
 
 # *** ComputationHandle ***
 
-
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2355]: two empty lines

File: packages/python/accern_xyme/accern_xyme.py (hunk 11 above)
Comment (JosuaKrause) [reply packages/python/accern_xyme/accern_xyme.py:2404]: double check

File: packages/typescript/index.ts [hunk 13]
```diff
@@ -1642,17 +1685,24 @@ export class NodeHandle {
...
     ) {
         this.client = client;
         this.dag = dag;
-        this.id = id;
+        this.nodeId = nodeId;
         this.name = name;
         this.type = kind;
     }
 
+    private asOwner(): BlobOwner {
```
Comment (JosuaKrause) [inline packages/typescript/index.ts:1699]: same as above -- don't assume this.asOwner

File: packages/typescript/index.ts [hunk 14]
```diff
@@ -1839,6 +1890,7 @@ export class NodeHandle {
             })
             .then((response) => response.times);
     }
+
```
Comment (JosuaKrause) [inline packages/typescript/index.ts:1893]: lint?

File: packages/typescript/index.ts [hunk 15]
```diff
@@ -1861,18 +1913,49 @@ export class NodeHandle {
         if (uri === undefined) {
             throw new Error('uri is undefined');
         }
-        return new BlobHandle(this.client, uri, true);
+        return new BlobHandle(this.client, uri, true, this.asOwner());
```
Comment (JosuaKrause) [inline packages/typescript/index.ts:1916]: no no no

File: packages/typescript/index.ts [hunk 16]
```diff
@@ -1904,39 +1987,27 @@ export class NodeHandle {
         return new CSVBlobHandle(this.client, res.csv, owner);
     }
 
+    // public async getJSONBlob(): Promise<JSONBlobHandle> {
```
Comment (JosuaKrause) [inline packages/typescript/index.ts:1990]: maybe remove?

File: packages/typescript/index.ts [hunk 17]
```diff
@@ -2037,7 +2099,10 @@ export class BlobHandle {
                 });
                 return interpretCtype(res, ctype);
             } catch (error) {
-                if (error.response.status === 404) {
+                if (
+                    error instanceof HTTPResponseError &&
```
Comment (JosuaKrause) [inline packages/typescript/index.ts:2103]: if !(error instanceof HTTPResponseError) throw error

File: packages/typescript/index.ts [hunk 18]
```diff
@@ -2115,7 +2180,11 @@ export class BlobHandle {
                 to_uri: toUri,
             },
         });
```
Comment (JosuaKrause) [inline packages/typescript/index.ts:2182]: how would this work with external owner?

File: packages/typescript/index.ts [hunk 19]
```diff
@@ -2357,26 +2426,14 @@ export class BlobHandle {
...
-                blob: this.getUri(),
-            },
-        });
-    }
 }
 
 export class CSVBlobHandle extends BlobHandle {
-    owner: BlobOwner;
-
     constructor(client: XYMEClient, uri: string, owner: BlobOwner) {
```
Comment (JosuaKrause) [inline packages/typescript/index.ts:2435]: constructor necessary?

File: setup.cfg [hunk 20]
```diff
@@ -24,7 +24,7 @@ package_dir =
 packages = find:
 python_requires = >=3.6
 install_requires =
-    # dvc>=2.0.17
+    dvc>=2.0.17
```
Comment (JosuaKrause) [inline setup.cfg:27]: remove from here though

File: packages/python/accern_xyme/accern_xyme.py [hunk 21]
```diff
@@ -834,6 +866,33 @@ def get_triton_models(self) -> List[str]:
...
+                Git commit (any revision such as a branch or tag name, or a
+                commit hash). If repo is not a Git repo, this option is
+                ignored. Default: HEAD.
+
+        Returns:
+            the content of the file.
+        """
+        import dvc.api
+
+        res = dvc.api.read(path, repo=repo, rev=rev, mode="r")
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:889]: with ?

File: packages/python/accern_xyme/accern_xyme.py [hunk 22]
```diff
@@ -2323,41 +2402,116 @@ def add_from_df(
                 io_in.close()
             self._clear_upload()
 
+
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2405]: no, that belongs to the above class -- no extra line here

File: packages/python/accern_xyme/accern_xyme.py [hunk 23]
```diff
@@ -2323,41 +2402,116 @@ def add_from_df(
...
         }
         if requeue_on_finish is not None:
             obj["dag"] = requeue_on_finish.get_dag().get_uri()
             obj["node"] = requeue_on_finish.get_id()
-        res = cast(JSONBlobAppendResponse, self._client.request_json(
-            METHOD_PUT, "/json_append", obj))
-        self._count = res["count"]
-        return self
+        self._client.request_json(METHOD_PUT, "/json_append", obj)
+
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2427]: no extra line

File: packages/python/accern_xyme/accern_xyme.py [hunk 24]
```diff
@@ -2323,41 +2402,116 @@ def add_from_df(
...
+            }))
+
+    def get_custom_code(self) -> NodeCustomCode:
+        return cast(NodeCustomCode, self._client.request_json(
+            METHOD_GET, "/custom_code", {
+                "dag": self.get_owner_dag(),
+                "node": self.get_owner_node(),
+            }))
+
+
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2481]: no extra line

File: packages/python/accern_xyme/accern_xyme.py [hunk 25]
```diff
@@ -2323,41 +2402,116 @@ def add_from_df(
...
+                "config": obj,
+            }))
+
+    def get_model_params(self) -> ModelParamsResponse:
+        return cast(ModelParamsResponse, self._client.request_json(
+            METHOD_GET, "/model_params", {
+                "dag": self.get_owner_dag(),
+                "node": self.get_owner_node(),
+            }))
+
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/accern_xyme.py:2510]: no extra line

File: packages/python/accern_xyme/types.py [hunk 26]
```diff
@@ -252,13 +241,7 @@
...
     "modules": List[str],
 })
-JSONBlobResponse = TypedDict('JSONBlobResponse', {
-    "count": int,
-    "json": str,
-})
-JSONBlobAppendResponse = TypedDict('JSONBlobAppendResponse', {
-    "count": int,
-})
+
```
Comment (JosuaKrause) [inline packages/python/accern_xyme/types.py:244]: no empty line?

File: packages/python/accern_xyme/accern_xyme.py [hunk 27]
```diff
@@ -1627,6 +1726,12 @@ def __init__(
         self._state: Optional[int] = None
         self._config_error: Optional[str] = None
 
+    def as_owner(self) -> BlobOwner:
```
Comment (jckhang) [inline packages/python/accern_xyme/accern_xyme.py:1729]: remove

File: packages/typescript/index.ts [hunk 28]
```diff
@@ -199,59 +206,196 @@ export default class XYMEClient {
...
+                    ) {
+                        retry = MAX_RETRY;
+                    }
+                    throw e;
+                }
+            } catch (error) {
+                if (retry >= MAX_RETRY) {
+                    throw error;
+                }
+                if (method in NO_RETRY) {
```
Comment (jckhang) [inline packages/typescript/index.ts:244]: indexOf

File: packages/typescript/index.ts [hunk 29]
```diff
@@ -199,59 +206,196 @@ export default class XYMEClient {
...
```
Comment (jckhang) [inline packages/typescript/index.ts:290]: indexof

File: packages/typescript/index.ts [hunk 30]
```diff
@@ -199,59 +206,196 @@ export default class XYMEClient {
...
```
Comment (jckhang) [inline packages/typescript/index.ts:336]: indexof

File: packages/typescript/index.ts [hunk 31]
```diff
@@ -199,59 +206,196 @@ export default class XYMEClient {
...
                 break;
             }
             case METHOD_POST:
             case METHOD_PUT:
             case METHOD_DELETE: {
-                response = await retryWithTimeout(url, retry, options);
+                response = await fetch(url, {
+                    method: METHOD_POST,
+                    headers: {
+                        'Authorization': headers['Authorization'],
```
Comment (jckhang) [inline packages/typescript/index.ts:391]: authorization

Comment (jckhang) [reply packages/typescript/index.ts:391]: ...headers,

File: packages/typescript/index.ts [hunk 32]
```diff
@@ -375,53 +515,40 @@ export default class XYMEClient {
...
             };
         }
         const options: RequestInit = {
-            ...rest,
             method,
-            headers,
-            body: JSON.stringify(args),
+            headers: {
+                ...headers,
+                'content-type': 'application/text',
```
Comment (jckhang) [inline packages/typescript/index.ts:543]: 'application/json'

File: packages/typescript/index.ts [hunk 33]
```diff
@@ -549,17 +671,20 @@ export default class XYMEClient {
...
+        const dag = new DagHandle(this, dagURI);
+        this.dagCache.set({ uri: dagURI }, dag);
         return dag;
     }
 
-    public getBlobHandle(uri: string, isFull = false): BlobHandle {
+    public async getBlobHandle(
+        uri: string,
+        isFull = false
+    ): Promise<BlobHandle> {
```
Comment (jckhang) [inline packages/typescript/index.ts:687]: no promise

File: packages/typescript/index.ts [hunk 34]
```diff
@@ -2495,23 +2686,135 @@ export class CSVBlobHandle extends BlobHandle {
...
-                owner_node: this.owner.owner_node,
+                owner_dag: this.getOwnerDag(),
+                owner_node: this.getOwnerNode(),
+            },
+        });
+    }
+}
+
+export class CustomCodeBlobHandle extends BlobHandle {
+    validBlobTypes = ['custom_code'];
```
Comment (jckhang) [inline packages/typescript/index.ts:2707]: empty line

**Review by JosuaKrause — APPROVED** [review]
lgtm
